packages/remix-server-runtime/injectServerEffect.ts

```diff
@@ -0,0 +1,133 @@
+export type ServerEffectFunction<Args extends Array<unknown>> = (
+  ...args: Args
+) => Promise<ServerCleanupFunction> | ServerCleanupFunction;
+
+export type ServerCleanupFunction = () => Promise<void> | void;
+
+export type ServerEffectsMap = Map<string, ServerEffect<Array<unknown>>>;
+
+export const SERVER_EFFECTS_KEY = Symbol("remixServerEffects");
+
+/**
+ * Creates a server-side effect persisted on tihs Node.js process.
+ * Cleans up and re-evaluates the effect on HMR and HDR.
+ * @param id Unique string ID of the effect. Effects are persisted
+ * and deduplicated by their IDs.
+ * @param effect The effect function.
+ * @param dependencies The list of effect arguments (dependencies).
+ * Pass the arguments to the `effect` in the dependencies array so
+ * the effect gets refreshes on HMR and HDR.
+ *
+ * @example
+ * import { init } from 'third-party'
+ * import { payload } from './my/utils.server'
+ *
+ * injectServerEffect('myEffect', (payload) => {
+ *  init(payload)
+ * }, [payload])
+ */
+export async function injectServerEffect<Dependencies extends Array<unknown>>(
+  id: string,
+  callback: ServerEffectFunction<Dependencies>,
+  dependencies: Dependencies
+): Promise<ServerEffect<Dependencies>> {
+  let lastEffect = getGlobalEffectById(id);
+
+  if (typeof lastEffect !== "undefined") {
+    // First, clean up the last effect.
+    await lastEffect.dispose();
+
+    // Then, re-run the effect, updating its result
+    // and effect entry methods in the global reference.
+    await lastEffect.run(...dependencies);
+
+    return lastEffect;
+  }
+
+  // Create a new effect and run it.
+  let effect = new ServerEffect(id, callback);
+  await effect.run(...dependencies);
+
+  // Whenever the process is terminated, remove the effect.
+  process.on("SIGTERM", effect.dispose).on("SIGINT", effect.dispose);
+
+  return effect;
+}
+
+export class ServerEffect<Dependencies extends Array<unknown>> {
```

Comment:
Would it be possible to use normal objects + functions instead?
The only classes we currently use in the codebase are for creating custom `Error`s

Reply:
I don't mind. My motivation for using a class was to manage that self-updating `this.cleanup` reference. I can use the functional approach although, imo, it's a bit harder to understand than the class.

Reply:
Let's remove the class for consistency with the rest of the codebase (where we don't use classes)